Sources/Purchasing/Purchases/Purchases.swift

```diff
@@ -962,13 +966,47 @@ public extension Purchases {
 
     @objc(purchasePackage:withCompletion:)
     func purchase(package: Package, completion: @escaping PurchaseCompletedBlock) {
-        purchasesOrchestrator.purchase(product: package.storeProduct, package: package, completion: completion)
+        purchasesOrchestrator.purchase(product: package.storeProduct,
+                                       package: package,
+                                       promotionalOffer: nil,
+                                       metadata: nil,
+                                       completion: completion)
     }
 
     func purchase(package: Package) async throws -> PurchaseResultData {
         return try await purchaseAsync(package: package)
     }
 
+    #if ENABLE_PURCHASE_PARAMS
+
+    @objc(purchaseProduct:params:withCompletion:)
+    func purchase(product: StoreProduct, params: PurchaseParams, completion: @escaping PurchaseCompletedBlock) {
+        purchasesOrchestrator.purchase(product: product,
+                                       package: nil,
+                                       promotionalOffer: params.promotionalOffer?.signedData,
+                                       metadata: params.metadata,
+                                       completion: completion)
+    }
+
+    func purchase(product: StoreProduct, params: PurchaseParams) async throws -> PurchaseResultData {
+        return try await purchaseAsync(product: product, params: params)
+    }
+
+    @objc(purchasePackage:params:withCompletion:)
+    func purchase(package: Package, params: PurchaseParams, completion: @escaping PurchaseCompletedBlock) {
+        purchasesOrchestrator.purchase(product: package.storeProduct,
```

Comment:
I suppose this would be the best place to switch to the different purchase() functions based on the present subscription offer once we support different offer types in the purchase params?

Just thinking ahead to when we support win-back offers :D

Reply:
I think we can just add an extra optional parameter to the existing purchase methods in PurchasesOrchestrator like I did here for metadata:

https://github.com/RevenueCat/purchases-ios/pull/4400/files#diff-c6164c2ee507c30293a823c9e66c5951f7dbed017e61a1115aaaf6e95a57d8a5R488